builtin/logical/totp/path_code.go

```diff
@@ -0,0 +1,110 @@
+package totp
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	totplib "github.com/pquerna/otp/totp"
+)
+
+func pathCode(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "code/" + framework.GenericNameRegex("name"),
+		Fields: map[string]*framework.FieldSchema{
+			"name": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Name of the key.",
+			},
+			"code": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "TOTP code to be validated.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathReadCode,
+			logical.UpdateOperation: b.pathValidateCode,
+		},
+
+		HelpSynopsis:    pathCodeHelpSyn,
+		HelpDescription: pathCodeHelpDesc,
+	}
+}
+
+func (b *backend) pathReadCode(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+
+	// Get the key
+	key, err := b.Key(req.Storage, name)
+	if err != nil {
+		return nil, err
+	}
+	if key == nil {
+		return logical.ErrorResponse(fmt.Sprintf("unknown key: %s", name)), nil
+	}
+
+	// Generate password using totp library
+	totpToken, err := totplib.GenerateCodeCustom(key.Key, time.Now(), totplib.ValidateOpts{
+		Period:    key.Period,
+		Digits:    key.Digits,
+		Algorithm: key.Algorithm,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the secret
+	resp, err := &logical.Response{
```

Comment:
s/resp, err :=/return/ ? There's really no reason to define both of these only to return them, just return them directly.